### PR TITLE
Add a separate collection for related sheets for Formula plugin

### PR DESCRIPTION
**Pull request by budnix**
### Context

The PR fixes an error that occurs when the HF engine instance was passed to the _Formulas_ plugin configuration. For that kind of configuration, the collection that holds the Handsontable instances was not created properly. Hence, the table that depends on the changed formulas could not be updated.

To solve that issue and with support passing HF instances and Hyperformula classes in mind, I added a separate collection (based on static registered module) for connecting HoT instances with the engine. Previously the collection was mixed with `sharedEngineUsage` but that collection has a different meaning.

Moreover, I found a memory leak [(missing `unregisterEngine` call)](https://github.com/handsontable/handsontable/pull/8100/files#diff-75a3f65e47cad775c4d9cef1c98b6ef3ad0ad696c9aaa6b62eaf9fe45a09833cR165) that occurs while calling the `updateSettings` method. I fixed that and cover the changes with the tests.

Demo to reproduce:
```js
const hyperformulaInstance = HyperFormula.buildEmpty();

const hot1 = new Handsontable(document.getElementById('example'), {
  data: [
    ['10.26', '', 'Sum', '=SUM(A:A)'],
    ['20.12', '', 'Average', '=AVERAGE(A:A)'],
    ['30.01', '', 'Median', '=MEDIAN(A:A)'],
    ['40.29', '', 'MAX', '=MAX(A:A)'],
    ['50.18', '', 'MIN', '=MIN(A:A)'],
  ],
  colHeaders: true,
  rowHeaders: true,
  formulas: {
    engine: hyperformulaInstance,
    sheetName: 'Sheet1',
  },
});
const hot2 = new Handsontable(document.getElementById('example2'), {
  data: [
    ['Is A1 in Sheet1 > 10?', '=IF(Sheet1!A1>10,"TRUE","FALSE")'],
    ['Is A:A in Sheet > 150?', '=IF(SUM(Sheet1!A:A)>150,"TRUE","FALSE")'],
    ['How many blank cells are in the Sheet1?', '=COUNTBLANK(Sheet1!A1:E5)'],
    ['Generate a random number', '=RAND()'],
    ['Number of sheets in this workbook', '=SHEETS()'],
  ],
  colHeaders: true,
  rowHeaders: true,
  formulas: {
    engine: hyperformulaInstance,
    sheetName: 'Sheet2',
  },
});
```
While editing the Sheet1 the error occurred.

_[skip changelog]_

### How has this been tested?

I added new E2E tests.

### Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature or improvement (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Additional language file or change to the existing one (translations)

### Related issue(s):
1. #8097
2.
3.

### Affected project(s):
- [x] `handsontable`
- [x] `@handsontable/angular`
- [x] `@handsontable/react`
- [x] `@handsontable/vue`

### Checklist:

- [x] I have reviewed the guidelines about [Contributing to Handsontable](https://github.com/handsontable/handsontable/blob/master/CONTRIBUTING.md) and I confirm that my code follows the code style of this project.
- [x] I have signed the [Contributor License Agreement](https://docs.google.com/forms/d/e/1FAIpQLScpMq4swMelvw3-onxC8Jl29m0fVp5hpf7d1yQVklqVjGjWGA/viewform?c=0&w=1)
- [ ] My change requires a change to the documentation.

### CODE REVIEW

File: src/plugins/formulas/engine/register.js [hunk 1]
```diff
@@ -4,7 +4,37 @@ import { warn } from '../../../helpers/console';
...
+function getSharedEngineUsageRegistry() {
+  const registryKey = 'shared_engine_usage';
+  const pluginStaticRegistry = staticRegister(PLUGIN_KEY);
+
+  if (!pluginStaticRegistry.hasItem(registryKey)) {
+    pluginStaticRegistry.register(registryKey, new Map());
+  }
+
+  return pluginStaticRegistry.getItem(registryKey);
+}
```
Comment (wojciechczerniak) [inline src/plugins/formulas/engine/register.js:37]: Looks like we can deduplicate the code for `getEngineRelationshipRegistry` and `getSharedEngineUsageRegistry`
```suggestion
function getSharedEngineUsageRegistry() {
  return getEngineUsageRegistry('shared_engine_usage');
}

function getEngineUsageRegistry(registryKey) {
  const pluginStaticRegistry = staticRegister(PLUGIN_KEY);

  if (!pluginStaticRegistry.hasItem(registryKey)) {
    pluginStaticRegistry.register(registryKey, new Map());
  }

  return pluginStaticRegistry.getItem(registryKey);
}
```

File: src/plugins/formulas/engine/register.js [hunk 2]
```diff
@@ -118,14 +148,24 @@ export function getRegisteredHotInstances(engine) {
...
+    }
+
+    if (sharedEngineUsage && sharedEngineUsage.includes(hotInstance.guid)) {
+      sharedEngineUsage.splice(sharedEngineUsage.indexOf(hotInstance.guid), 1);
+
+      if (sharedEngineUsage.length === 0) {
+        sharedEngineRegistry.delete(engine);
         engine.destroy();
       }
     }
```
Comment (wojciechczerniak) [inline src/plugins/formulas/engine/register.js:171]: We can probably deduplicate those loops by extracting `deleteEngine` function 🤔

**Review by wojciechczerniak — APPROVED** [review]
👍

**Review by aninde — APPROVED** [review]
I did a few demo tests, I couldn't break your solution, well done 👍

File: src/plugins/formulas/engine/register.js (hunk 2 above)
Comment (budnix) [reply src/plugins/formulas/engine/register.js:171]: I'll leave it for the next iteration.

**Pull request merged**